core/src/main/java/org/apache/calcite/rel/rules/DateRangeRules.java

```diff
@@ -401,21 +402,33 @@ private RexNode toRex(RexNode operand, Range<Calendar> r) {
             : SqlStdOperatorTable.GREATER_THAN;
         nodes.add(
             rexBuilder.makeCall(op, operand,
-                rexBuilder.makeDateLiteral(
-                    DateString.fromCalendarFields(r.lowerEndpoint()))));
+                makeLiteralForComparison(rexBuilder, r.lowerEndpoint(), operand)));
       }
       if (r.hasUpperBound()) {
         final SqlBinaryOperator op = r.upperBoundType() == BoundType.CLOSED
             ? SqlStdOperatorTable.LESS_THAN_OR_EQUAL
             : SqlStdOperatorTable.LESS_THAN;
         nodes.add(
             rexBuilder.makeCall(op, operand,
-                rexBuilder.makeDateLiteral(
-                    DateString.fromCalendarFields(r.upperEndpoint()))));
+                makeLiteralForComparison(rexBuilder, r.upperEndpoint(), operand)));
       }
       return RexUtil.composeConjunction(rexBuilder, nodes, false);
     }
 
+    private RexLiteral makeLiteralForComparison(RexBuilder rexBuilder, Calendar calendar,
+        RexNode operand) {
+      switch (operand.getType().getSqlTypeName()) {
+      case TIMESTAMP:
+      case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
```

Comment:
don't we need the time zone field to be part of literal?

Reply:
I don't think so. A TIMESTAMP_WITH_LOCAL_TIME_ZONE is always interpreted in local time, so the TZ is implicit.